providers/base/bin/gateway_ping_test.py

```diff
@@ -83,151 +108,297 @@ def _get_default_gateway_from_bin_route(self):
-            _("Reading default gateway information from route binary"))
-        routebin = subprocess.getstatusoutput(
-            "export LANGUAGE=C; " "/usr/bin/env route -n")
-        if routebin[0] == 0:
-            h = re.compile(r"\n0.0.0.0\s+(?P<def_gateway>[\w.]+)\s+")
-            w = h.search(routebin[1])
-            if w:
-                def_gateway = w.group("def_gateway")
-                if def_gateway:
-                    return def_gateway
+            _("Reading default gateway information from route binary")
+        )
+        try:
+            routebin = subprocess.check_output(
+                ["/usr/bin/env", "route", "-n"],
+                env={"LANGUAGE": "C"},
+                universal_newlines=True,
+            )
+        except subprocess.CalledProcessError:
+            return None
+        route_line_re = re.compile(
+            r"^0\.0\.0\.0\s+(?P<def_gateway>[\w.]+)(?P<tail>.+)",
+            flags=re.MULTILINE,
+        )
+        route_lines = route_line_re.finditer(routebin)
+        for route_line in route_lines:
+            def_gateway = route_line.group("def_gateway")
+            interface = route_line.group("tail").rsplit(" ", 1)[-1]
+            if interface == self.interface and def_gateway:
+                return def_gateway
         logging.error(_("Could not find default gateway by running route"))
         return None
 
-    def get_hostname(self):
-        return socket.gethostname()
-
-    def get_default_gateway(self):
-        t1 = self._get_default_gateway_from_proc()
-        if not t1:
-            t1 = self._get_default_gateway_from_bin_route()
-        return t1
-
-
-def get_host_to_ping(interface=None, verbose=False, default=None):
-    # Get list of all IPs from all my interfaces,
-    interface_list = subprocess.check_output(["ip", "-o", 'addr', 'show'])
-    reg = re.compile(r'\d: (?P<iface>\w+) +inet (?P<address>[\d\.]+)/'
-                     r'(?P<netmask>[\d]+) brd (?P<broadcast>[\d\.]+)')
-    # Will magically exclude lo because it lacks brd field
-    interfaces = reg.findall(interface_list.decode())
-    # ping -b the network on each one (one ping only)
-    # exclude the ones not specified in iface
-    for iface in interfaces:
-        if not interface or iface[0] == interface:
-            # Use check_output even if I'll discard the output
-            # looks cleaner than using .call and redirecting stdout to null
-            try:
-                subprocess.check_output(["ping", "-q", "-c", "1", "-b",
-                                         iface[3]], stderr=subprocess.STDOUT)
-            except subprocess.CalledProcessError:
-                pass
-    # If default host given, ping it as well,
-    # to try to get it into the arp table.
-    # Needed in case it's not responding to broadcasts.
-    if default:
+    def _get_ip_addr_info(self):
+        return subprocess.check_output(
+            ["ip", "-o", "addr", "show"], universal_newlines=True
+        )
```

Comment:
I wouldnt add the decorator, there is value to explicitly checking the exception locally (namely, you can not accidentally non-raise a second command that is supposed to fail if you ever add it).

I didn't catch the exception because I don't expect it to ever fail. Do you think I should? I think that if `ip addr show` fails, the script should blow up.

Reply:
Nah, It is okey like that, I just wanted to know your opinion about it